### ISSUE TITLE
gpu-next: using VK_KHR_cooperative_matrix extension

### ISSUE THREAD

**Issue by arm64-v9a** [mpv-player/mpv#12144]
Before requesting a new feature make sure it hasn't been requested yet.
https://github.com/mpv-player/mpv/labels/meta%3Afeature-request

### Expected behavior of the wanted feature
Support using `VK_KHR_cooperative_matrix` extension

https://github.com/bloc97/Anime4K/issues/145#issuecomment-1509999595

It may be able to improve the performance of shaders such as Anime4K and address the stagnation in development due to performance problem.

[NCNN](https://github.com/Tencent/ncnn/pull/4823)  and [FFmpeg](https://ffmpeg.org/pipermail/ffmpeg-devel/2023-August/313213.html)  have already been accelerated by this extension.
### Alternative behavior of the wanted feature
N/A

**Comment by bjin** [comment]
Just to be clear, adding support of this to mpv or libplacebo is the least important blocking issue. It could be as simple as adding a `"#EXTENSION"` directive to shader prelude (probably via `//!EXTENTION"`).

Utilizing this extension (in user shader), however, is quite complicated. This is especially true for CNN shaders like FSRCNNx and Anime4k. It basically means writing a new shader from scratch with 10x complexity (compute shader, subgroup, buffer storage, fp16 ...). And even if all these are done, There are different subgroup-size/coopMatMul kernel size available from different vendor implementation, and their performance will vary between different GPUs. Modern DL framework like pytorch and tensorflow will actually compare and choose different kernel at runtime for best performance.

So, instead of opening meaningless feature request here, you should probably go to those repos and open FR there.


side story:

I thought about using this extension in my nnedi3 shader, because it's a much simpler case: single layer and kernel size (8x4 and 8x6) happens to be multiple of 16 so no routines needed for leftovers. But it still requires a lot of effort, and probably too much for a somehow outdated model like nnedi3.

**Comment by bjin** [comment]
> But libplacebo's built-in shaders may still benefit from it.

No, those shaders won't. Fast matrix multiplication only benefits massive convolution kernel with large input depth (8 to be precise).

**Comment by haasn** [comment]
I would definitely be interested in a fast full image DCT implementation. Especially in combination with a `FREQUENCY`/`PHASE` hook stage to allow user shaders to apply arbitrary transformations to the image in the fourier domain.

Could potentially use this e.g. instead of cascading gaussian blurs for very large blur factors, and maybe for extreme downscaling (16x or more) which is prohibitively expensive with conventional convolution (especially when the convolution kernel is unrolled), obviously denoising, some types of film grain generation, full frame blue noise dithering, etc...
